### PR TITLE
Added 'openArray[char]' overloads to 'std/parseutils'

### PR DESCRIPTION
**Pull request by beef331**

### CODE REVIEW

**Comment by beef331** [comment]
Seems an issue will be converters to `string` do not fire for `openarray[char]`. Perhaps for converters `string` should work with `openArray[char]`?

**Comment by Araq** [comment]
> Seems an issue will be converters to string do not fire for openarray[char]. Perhaps for converters string should work with openArray[char]?

Nah, it's fine this way IMO. The real issue is here that it's a breaking change which would be easily avoidable by introducing duplication in the form of a new `std / parseslices` module.

**Comment by beef331** [comment]
In the case where people want a 'weak distinct' it would give the ability to optimise a copy away as the following. Given `openarray[char]` cannot be used in a converter one cannot really do this.
```nim
type NotString = distinct string
converter toString*(s: NotString): lent string = string(s)
```

**Comment by Araq** [comment]
I understand the problem but introducing "converter chaining" at this point would be far too an invasive change. So the next best option is `str / parseslices`

**Comment by Varriount** [comment]
This is a kludgy hack. Is there any alternative to creating a copy of an entire module? Why can't the duplicate functions just be added to `parseutils`? 

> I understand the problem but introducing "converter chaining" at this point would be far too an invasive change.

Even if it's only for special cases (like openarrays)?

**Comment by beef331** [comment]
> Why can't the duplicate functions just be added to parseutils

I forgot that `openarray[char]` and `string` did not cause a mismatch with converters, so I guess it could be added there, make all the base procedures operate on `openarray[char]` but have the `string` versions call the `openarray` variants.

**Comment by Araq** [comment]
> Even if it's only for special cases (like openarrays)?

*Especially* if it's only for special cases. These come back and bite you. ;-)

**Comment by planetis-m** [comment]
For custom string types it's nice to have implicit converters to openArray:

```nim
{.experimental: "views".}

converter toOpen(x: MyString): openArray[byte] =
  toOpenArray(x.data, 0, x.data.high)
```

although I don't know if this is the direction we are going, or is it going to cause problems?

**Comment by Varriount** [comment]
> although I don't know if this is the direction we are going, or is it going to cause problems?

The direction I'd like here is to solve this problem in a simple way, that minimizes redundancy. Ideally by just adding overloads and/or type constraints to procedures in parseutils.

**Comment by Araq** [comment]
Adding overloads that support openArray[char] seems to be the best option for now.

File: lib/system.nim [hunk 1]
```diff
@@ -2580,7 +2580,7 @@ template once*(body: untyped): untyped =
 
 {.pop.} # warning[GcMem]: off, warning[Uninit]: off
 
-proc substr*(s: string, first, last: int): string =
+proc substr*(s: openArray[char], first, last: int): string =
```
Comment (Araq) [inline lib/system.nim]: But thanks to `toOpenArray` the slice `first..last` idea is already part of the `openArray[char]` type. This is a very weird interface.

File: lib/pure/parseutils.nim [hunk 2]
```diff
@@ -242,7 +244,7 @@ proc parseIdent*(s: string, start = 0): string =
     while i < s.len and s[i] in IdentChars: inc(i)
     result = substr(s, start, i-1)
 
-proc parseChar*(s: string, c: var char, start = 0): int =
+proc parseChar*(s: openArray[char], c: var char, start = 0): int =
```
Comment (Araq) [inline lib/pure/parseutils.nim]: openArray allows for effective slicing making the `start` parameters redundant and weird.

File: lib/system.nim [hunk 3]
```diff
@@ -2601,9 +2601,15 @@ proc substr*(s: string, first, last: int): string =
   for i in 0 .. L-1:
     result[i] = s[i+first]
 
-proc substr*(s: string, first = 0): string =
+proc substr*(s: string, first, last: int): string =
+  result = substr(openArray[char](s), first, last)
+
+proc substr*(s: openArray[char], first = 0): string =
```
Comment (Araq) [inline lib/system.nim]: Likewise.

File: lib/system/strmantle.nim [hunk 4]
```diff
@@ -78,7 +78,7 @@ const
 when defined(nimHasInvariant):
   {.push staticBoundChecks: off.}
 
-proc nimParseBiggestFloat(s: string, number: var BiggestFloat,
+proc nimParseBiggestFloat(s: openArray[char], number: var BiggestFloat,
```
Comment (Araq) [inline lib/system/strmantle.nim:81]: Likewise.

File: lib/system.nim (hunk 1 above)
Comment (beef331) [reply lib/system.nim]: `openArray[char][a..b]` returns a `seq[char]` and not a `string`,  so in the cases `substr` is used one will likely need an equivalent regardless, perhaps it should be named `toString` instead for the `openarray` variant?

Comment (Araq) [reply lib/system.nim]: I didn't say `openArray[char][a..b]`, I said `toOpenArray`.

File: lib/system.nim [hunk 5]
```diff
@@ -2580,7 +2580,19 @@ template once*(body: untyped): untyped =
...
-proc substr*(s: string, first, last: int): string =
+proc substr*(s: openArray[char]): string =
+  ## Copies a slice of `s` into a new string and returns this new
+  ## string.
+  runnableExamples:
+    let a = "abcdefgh"
+    assert a.substr(2, 5) == "cdef"
+    assert a.substr(2) == "cdefgh"
+    assert a.substr(5, 99) == "fgh"
+  result = newString(s.high)
```
Comment (Araq) [inline lib/system.nim]: Shouldn't this be `s.high + 1`?

Comment (beef331) [reply lib/system.nim]: Correct

File: lib/pure/parseutils.nim [hunk 6]
```diff
@@ -59,7 +61,7 @@ const
 proc toLower(c: char): char {.inline.} =
   result = if c in {'A'..'Z'}: chr(ord(c)-ord('A')+ord('a')) else: c
 
-proc parseBin*[T: SomeInteger](s: string, number: var T, start = 0,
+proc parseBin*[T: SomeInteger](s: openArray[char], number: var T, start = 0,
```
Comment (Araq) [inline lib/pure/parseutils.nim]: as I said, what takes an openArray doesn't need a `start` parameter.

**Review by Varriount — APPROVED** [review]

**Comment by ZoomRmc** [comment]
Closes #14810

**Pull request merged**